src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs

```diff
@@ -0,0 +1,239 @@
+            if (accountNeth is null)
+            {
+                return (responseNodes.ToArray(), null);
+            }
+            Keccak? storageRoot = accountNeth.StorageRoot;
+
+            startingHash = startingHash == null ? Keccak.Zero : startingHash;
+            limitHash = limitHash == null ? Keccak.MaxValue : limitHash;
+
+            (Dictionary<byte[], byte[]>? requiredNodes, long innerResponseSize, bool stopped) = GetNodesFromTrieVisitor(storageRoot,
+                startingHash, limitHash, byteLimit - responseSize, HardResponseByteLimit - responseSize, true);
+
+            PathWithStorageSlot[] nodes = new PathWithStorageSlot[requiredNodes.Count];
+            int index = 0;
+            foreach (PathWithStorageSlot? result in requiredNodes.Select(res => new PathWithStorageSlot(new Keccak(res.Key), res.Value)))
+            {
+                nodes[index] = result;
+                index += 1;
+            }
+            responseNodes.Add(nodes);
+            if (stopped || startingHash != Keccak.Zero)
+            {
+                byte[][]? proofs = GenerateRangeProof(tree, startingHash, nodes[^1].Path, storageRoot);
+                return (responseNodes.ToArray(), proofs);
+            }
+            responseSize += innerResponseSize;
+        }
+        return (responseNodes.ToArray(), null);
+    }
+
+    private (Dictionary<byte[], byte[]>?, long, bool) GetNodesFromTrieVisitor(Keccak rootHash, Keccak startingHash, Keccak limitHash,
+        long byteLimit, long hardByteLimit, bool isStorage = false)
+    {
+        PatriciaTree tree = new(_store, _logManager);
+
+        RangeQueryVisitor visitor = new(startingHash.Bytes, limitHash.Bytes, !isStorage, byteLimit, hardByteLimit, HardResponseNodeLimit);
+        VisitingOptions opt = new() { ExpectAccounts = false, KeepTrackOfAbsolutePath = true };
+        tree.Accept(visitor, rootHash, opt);
+        (Dictionary<byte[], byte[]>? requiredNodes, long responseSize) = visitor.GetNodesAndSize();
+
+        return (requiredNodes, responseSize, visitor._isStoppedDueToHardLimit);
+    }
+
+    private Account? GetAccountByPath(StateTree tree, Keccak rootHash, byte[] accountPath)
+    {
+        _logger.Info("GetAccountByPath");
+        byte[]? bytes = tree.Get(accountPath, rootHash);
+        return bytes is null ? null : _decoder.Decode(bytes.AsRlpStream());
+    }
+
+    private byte[][] GenerateRangeProof(PatriciaTree tree, Keccak start, Keccak end, Keccak rootHash)
+    {
+        VisitingOptions opt = new() { ExpectAccounts = false };
+        ProofCollector accountProofCollector = new(start.Bytes);
+        tree.Accept(accountProofCollector, rootHash, opt);
+        byte[][]? firstProof = accountProofCollector.BuildResult();
+
+        accountProofCollector = new ProofCollector(end.Bytes);
+        tree.Accept(accountProofCollector, rootHash, opt);
+        byte[][]? lastProof = accountProofCollector.BuildResult();
+
+        HashSet<byte[]> proofs = new();
```

Comment:
HashSet seems excessive here?

Reply:
they are not two values, there are two sets of multiple values, firstProof and lastProof contains multiple proofs and there are some common proofs in both the sets. Is there a better way to do this insted of using HashSet?